src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchWriter.cs

```diff
@@ -0,0 +1,569 @@
+
+            // Reset the cache of content IDs here. As per spec, content IDs are only unique inside a change set.
+            this.PayloadUriConverter.Reset();
+            this.CurrentOperationContentId = null;
+        }
+
+        /// <summary>
+        /// Creates an <see cref="ODataBatchOperationResponseMessage"/> for writing an operation of a batch response - implementation of the actual functionality.
+        /// </summary>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        /// <returns>The message that can be used to write the response operation.</returns>
+        protected override ODataBatchOperationResponseMessage CreateOperationResponseMessageImplementation(string contentId)
+        {
+            this.WritePendingMessageData(true);
+
+            // In responses we don't need to use our batch URL resolver, since there are no cross referencing URLs
+            // so use the URL resolver from the batch message instead.
+            this.CurrentOperationResponseMessage = ODataBatchOperationResponseMessage.CreateWriteMessage(
+                this.RawOutputContext.OutputStream,
+                /*operationListener*/ this,
+                this.PayloadUriConverter.BatchMessagePayloadUriConverter,
+                this.Container);
+            this.SetState(BatchWriterState.OperationCreated);
+
+            Debug.Assert(this.CurrentOperationContentId == null, "The Content-ID header is only supported in request messages.");
+
+            // write the operation's start boundary string
+            this.WriteStartBoundaryForOperation();
+
+            // write the headers and request separator line
+            ODataBatchWriterUtils.WriteResponsePreamble(this.RawOutputContext.TextWriter, changeSetBoundary != null, contentId);
+
+            return this.CurrentOperationResponseMessage;
+        }
+
+        /// <summary>
+        /// Writes all the pending headers and prepares the writer to write a content of the operation.
+        /// </summary>
+        protected override void StartBatchOperationContent()
+        {
+            Debug.Assert(this.CurrentOperationMessage != null, "Expected non-null operation message!");
+            Debug.Assert(this.RawOutputContext.TextWriter != null, "Must have a batch writer!");
+
+            // write the pending headers (if any)
+            this.WritePendingMessageData(false);
+
+            // flush the text writer to make sure all buffers of the text writer
+            // are flushed to the underlying async stream
+            this.RawOutputContext.TextWriter.Flush();
+        }
+
+        /// <summary>
+        /// Disposes the batch writer and set the 'OperationStreamRequested' batch writer state;
+        /// called after the flush operation(s) have completed.
+        /// </summary>
+        protected override void DisposeBatchWriterAndSetContentStreamRequestedState()

```

Comment:
>protected override void DisposeBatchWriterAndSetContentStreamRequestedState() [](start = 8, length = 77)

This is only called by the derived class and doesn't need to be in the base class. We should move it to multipart/mixed derived class.

Reply:
cleaned up. Thanks.

---
In reply to: [141119001](https://github.com/OData/odata.net/pull/960#discussion_r141119001) [](ancestors = 141119001)